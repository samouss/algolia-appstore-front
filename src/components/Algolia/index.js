export { default as Provider } from './Provider';
export { default as withInfiniteHits } from './withInfiniteHits';
export { default as withQuery } from './withQuery';
export * from './utils';
