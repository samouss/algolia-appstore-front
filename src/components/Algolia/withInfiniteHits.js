import React, { Component } from 'react';
import { getDisplayName } from 'core/utils';
import { ContextTypes } from './Provider';

const withInfiniteHits = ({
  hitsPerPage = 25,
} = {}) => WrappedComponent => {
  class WithInfiniteHits extends Component {

    constructor(props) {
      super(props);

      this.state = {
        hits: [],
        nbHits: 0,
        nbPages: 0,
        page: 0,
        processingTimeMS: 0,
        isLoading: true,
        isInitialLoad: true,
        isEndReached: false,
      };

      this.onNextPage = this.onNextPage.bind(this);
      this.updateState = this.updateState.bind(this);
    }

    componentDidMount() {
      const { algoliaHelper } = this.context;

      algoliaHelper.setQueryParameter(
        'hitsPerPage',
        hitsPerPage,
      );

      algoliaHelper.on('result', this.updateState);
    }

    componentWillUnmount() {
      this.context.algoliaHelper.removeListener('result', this.updateState);
    }

    onNextPage() {
      const { isLoading, isEndReached } = this.state;

      if (isLoading || isEndReached) {
        return;
      }

      this.setState(() => ({
        isLoading: true,
      }));

      this.context.algoliaHelper
        .nextPage()
        .search();
    }

    updateState(content) {
      const isResetResults = content.page === 0;
      const isEndReached = content.page === content.nbPages - 1;

      this.setState(prevState => ({
        hits: isResetResults ? content.hits
          : prevState.hits.concat(content.hits),
        nbHits: content.nbHits,
        nbPages: content.nbPages,
        page: content.page,
        processingTimeMS: content.processingTimeMS,
        isLoading: false,
        isInitialLoad: false,
        isEndReached,
      }));
    }

    render() {
      return (
        <WrappedComponent
          {...this.props}
          {...this.state}
          onNextPage={this.onNextPage}
        />
      );
    }

  }

  WithInfiniteHits.displayName = getDisplayName(
    WrappedComponent,
    'withInfiniteHits',
  );

  WithInfiniteHits.contextTypes = ContextTypes;

  return WithInfiniteHits;
};

export default withInfiniteHits;
