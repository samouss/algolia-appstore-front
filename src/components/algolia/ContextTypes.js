import PropTypes from 'prop-types';

export default {
  algoliaClient: PropTypes.object.isRequired,
  algoliaHelper: PropTypes.object.isRequired,
};
