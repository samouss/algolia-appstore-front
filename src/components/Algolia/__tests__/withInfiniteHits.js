import React from 'react';
import PropTypes from 'prop-types';
import { shallow } from 'enzyme';
import { createMockAlgoliaClient, createMockAlgoliaHelper } from 'test/algolia';
import withInfiniteHits from '../withInfiniteHits';

describe('algolia', () => {
  describe('withInfiniteHits', () => {
    const createContext = () => ({
      algoliaHelper: createMockAlgoliaHelper(),
      algoliaClient: createMockAlgoliaClient(),
    });

    const Component = ({ children, className }) => (
      <div className={className}>{children}</div>
    );

    Component.propTypes = {
      className: PropTypes.string.isRequired,
      children: PropTypes.node.isRequired,
    };

    Component.defaultProps = {
      className: '',
    };

    it('expect to render', () => {
      const context = createContext();

      const props = {
        className: 'sample-class-name',
      };

      const ApplyComponent = withInfiniteHits()(Component);

      const component = shallow(
        <ApplyComponent
          {...props}
        >
          <div>Content</div>
        </ApplyComponent>,
        { context },
      );

      expect(component).toMatchSnapshot();
    });

    describe('componentDidMount', () => {
      it('expect to call setQueryParameter with default parameter', () => {
        const context = createContext();
        const ApplyComponent = withInfiniteHits()(Component);

        const component = shallow(
          <ApplyComponent>
            <div>Content</div>
          </ApplyComponent>,
          { context },
        );

        component.instance().componentDidMount();

        expect(context.algoliaHelper.setQueryParameter).toHaveBeenCalledWith(
          'hitsPerPage',
          25,
        );
      });

      it('expect to call setQueryParameter with given parameter', () => {
        const context = createContext();
        const ApplyComponent = withInfiniteHits({
          hitsPerPage: 50,
        })(Component);

        const component = shallow(
          <ApplyComponent>
            <div>Content</div>
          </ApplyComponent>,
          { context },
        );

        component.instance().componentDidMount();

        expect(context.algoliaHelper.setQueryParameter).toHaveBeenCalledWith(
          'hitsPerPage',
          50,
        );
      });

      it('expect to subscribe to result event', () => {
        const context = createContext();
        const ApplyComponent = withInfiniteHits()(Component);

        const component = shallow(
          <ApplyComponent>
            <div>Content</div>
          </ApplyComponent>,
          { context },
        );

        component.instance().componentDidMount();

        expect(context.algoliaHelper.on).toHaveBeenCalledWith(
          'result',
          component.instance().updateState,
        );
      });
    });

    describe('componentWillUnmount', () => {
      it('expect to unsubscribe to result event', () => {
        const context = createContext();
        const ApplyComponent = withInfiniteHits()(Component);

        const component = shallow(
          <ApplyComponent>
            <div>Content</div>
          </ApplyComponent>,
          { context },
        );

        component.instance().componentWillUnmount();

        expect(context.algoliaHelper.removeListener).toHaveBeenCalledWith(
          'result',
          component.instance().updateState,
        );
      });
    });

    describe('onNextPage', () => {
      it('expect to call nextPage and search', () => {
        const context = createContext();
        const ApplyComponent = withInfiniteHits()(Component);

        const component = shallow(
          <ApplyComponent>
            <div>Content</div>
          </ApplyComponent>,
          { context },
        );

        // @NOTE: simulate initial load
        component.setState({ isLoading: false });

        component.instance().onNextPage();

        expect(component.state().isLoading).toBe(true);
        expect(component.state().isEndReached).toBe(false);
        expect(context.algoliaHelper.nextPage).toHaveBeenCalled();
        expect(context.algoliaHelper.search).toHaveBeenCalled();
      });

      it('expect to not call nextPage and search if loading', () => {
        const context = createContext();
        const ApplyComponent = withInfiniteHits()(Component);

        const component = shallow(
          <ApplyComponent>
            <div>Content</div>
          </ApplyComponent>,
          { context },
        );

        component.setState({
          isEndReached: true,
        });

        component.setState({ isLoading: true });

        component.instance().onNextPage();

        expect(context.algoliaHelper.nextPage).not.toHaveBeenCalled();
        expect(context.algoliaHelper.search).not.toHaveBeenCalled();
      });

      it('expect to not call nextPage and search if end reached', () => {
        const context = createContext();
        const ApplyComponent = withInfiniteHits()(Component);

        const component = shallow(
          <ApplyComponent>
            <div>Content</div>
          </ApplyComponent>,
          { context },
        );

        component.setState({
          isEndReached: true,
          isLoading: false,
        });

        component.instance().onNextPage();

        expect(context.algoliaHelper.nextPage).not.toHaveBeenCalled();
        expect(context.algoliaHelper.search).not.toHaveBeenCalled();
      });
    });

    describe('updateState', () => {
      it('expect to update with concatenate hits', () => {
        const context = createContext();
        const ApplyComponent = withInfiniteHits()(Component);

        const component = shallow(
          <ApplyComponent>
            <div>Content</div>
          </ApplyComponent>,
          { context },
        );

        const prevState = {
          hits: [
            { objectID: '123' },
            { objectID: '456' },
          ],
          nbHits: 1250,
          nbPages: 20,
          page: 5,
          processingTimeMS: 5,
        };

        const content = {
          hits: [
            { objectID: '789' },
            { objectID: '101' },
          ],
          nbHits: 1250,
          nbPages: 20,
          page: 6,
          processingTimeMS: 5,
        };

        const expectation = {
          hits: [
            { objectID: '123' },
            { objectID: '456' },
            { objectID: '789' },
            { objectID: '101' },
          ],
          nbHits: 1250,
          nbPages: 20,
          page: 6,
          processingTimeMS: 5,
          isLoading: false,
          isInitialLoad: false,
          isEndReached: false,
        };

        component.setState(prevState);

        component.instance().updateState(content);

        expect(component.state()).toEqual(expectation);
      });

      it('expect to update with hits', () => {
        const context = createContext();
        const ApplyComponent = withInfiniteHits()(Component);

        const component = shallow(
          <ApplyComponent>
            <div>Content</div>
          </ApplyComponent>,
          { context },
        );

        const prevState = {
          hits: [
            { objectID: '123' },
            { objectID: '456' },
          ],
          nbHits: 1250,
          nbPages: 20,
          page: 5,
          processingTimeMS: 5,
        };

        const content = {
          hits: [
            { objectID: '789' },
            { objectID: '101' },
          ],
          nbHits: 1250,
          nbPages: 20,
          page: 0,
          processingTimeMS: 5,
        };

        const expectation = {
          hits: [
            { objectID: '789' },
            { objectID: '101' },
          ],
          nbHits: 1250,
          nbPages: 20,
          page: 0,
          processingTimeMS: 5,
          isLoading: false,
          isInitialLoad: false,
          isEndReached: false,
        };

        component.setState(prevState);

        component.instance().updateState(content);

        expect(component.state()).toEqual(expectation);
      });
    });
  });
});
