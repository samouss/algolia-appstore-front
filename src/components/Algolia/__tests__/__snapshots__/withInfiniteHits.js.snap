// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`algolia withInfiniteHits expect to render 1`] = `
<Component
  className="sample-class-name"
  hits={Array []}
  isEndReached={false}
  isInitialLoad={true}
  isLoading={true}
  nbHits={0}
  nbPages={0}
  onNextPage={[Function]}
  page={0}
  processingTimeMS={0}
>
  <div>
    Content
  </div>
</Component>
`;
